InsertionSort/InsertionSort.cpp

```diff
@@ -1,25 +1,44 @@
 #include <iostream>
 #include <vector>
+#include <algorithm>
+#include <iterator>
 
-std::vector<int> InsertionSort(std::vector<int> input_vector) { 
-	int key = 0, i = 0;
-	for(int j = 2; j < input_vector.size(); ++j) {
-		key = input_vector[j];
+template <typename RandAccessContainer>
+RandAccessContainer &InsertionSort(RandAccessContainer &input_vector)
+{
+	int key{};
+	for (size_t j = 2; j < input_vector.size(); ++j)
+	{
+		// operator[] does not do range checking. Accessing element not presenting in vector silently leads to undefined behavior.
+		// at() member function does range checking and throws an exception when you are trying to access nonexisting element
+		key = input_vector.at(j);
+		// I changed the location of i declaration and its type, so as j, to eliminate not-safe conversions between i and j
+		// to signed types
+		static size_t i{}; // I made i static of this for loop in order to prevent declaring it over and over again in each iteration, so preserving its value
 		i = j - 1;
-		while (i > 0 && input_vector[i] > key) {
-			input_vector[i + 1] = input_vector[i];
-			i = i -1;
+		while (i > 0 && input_vector.at(i) > key)
+		{
+			input_vector.at(i + 1) = input_vector.at(i);
+			i = i - 1;
```

Comment:
In general, `[]` is faster than `.at()` when accessing elements in a vector. This is because `[]` does not perform bounds checking, so there is no additional overhead associated with checking if the index is within bounds.

On the other hand, `.at()` incurs the overhead of bounds checking, which involves verifying if the index is within the valid range and potentially throwing an exception if it is not. This extra checking can make `.at()` slightly slower compared to `[]`.

Therefore, if you are certain that the index is within the bounds of the vector, using `[]` is preferred for better performance. However, if you want the safety of bounds checking and are willing to trade off a bit of performance for it, then you can use `.at()`.